File: Blackjack21-JAVA/Blackjack21Library/src/Blackjack21/Game/Helper/ShuffleList.java
package Blackjack21.Game.Helper;

import java.util.ArrayList;
import java.util.Date;
import java.util.Random;

/**
 * A class used to shuffle a list's objects The object to use for the list
 *
 * @param <T> Object parameter
 */
public class ShuffleList<T> extends ArrayList<T>
{

    /**
     * Initialize the shuffle list object
     */
    public ShuffleList()
    {
    }

    /**
     * Add a new item to the list
     *
     * @param item Item to add to the list
     * @param isRandom Add the item to a random location
     */
    public void add(T item, boolean isRandom)
    {
        if (super.isEmpty() || isRandom == false)
        {
            super.add(item);
        }
        else
        {
            int index = next(0, size() + 1);
            if (index == size())
            {
                super.add(item);
            }
            else
            {
                T x = this.get(index);
                this.set(index, item);

                super.add(x);
            }
        }
    }

    /**
     * Shuffle items in the list by removing each item randomly and adding it to
     * another list before rebuilding the final list
     */
    public void shuffleShift()
    {
        ArrayList<T> tmp = new ArrayList<>();
        while (super.size() > 0)
        {
            int i = next(0, super.size());
            tmp.add(super.get(i));
            super.remove(i);

        }
        super.addAll(tmp);
        tmp.clear();
    }

    /**
     * Shuffle items in place by swapping locations randomly
     */
    public void shuffleInplace()
    {
        for (int i = super.size() - 1; i >= 0; i--)
        {
            T tmp = super.get(i);
            int randomIndex = next(0, i + 1);
            //Swap elements
            super.set(i, super.get(randomIndex));
            super.set(randomIndex, tmp);
        }
    }

    /**
     * Removes the first item on top of the list and returns it
     *
     * @return Returns the removed item
     */
    public T pop()
    {
        T t = super.get(0);
        super.remove(0);
        return t;
    }

    /**
     * Removes a random item from the list and returns it
     *
     * @return Returns the removed list item
     */
    public T popRandom()
    {
        int i = next(0, super.size());
        T t = super.get(i);
        super.remove(i);
        return t;
    }

    /**
     * Selects a random item from the list without removing it
     *
     * @return Returns the list item
     */
    public T selectRandom()
    {
        int i = next(0, super.size());
        T t = super.get(i);
        return t;
    }

    /**
     * Select the next random number
     *
     * @param min Minimum number
     * @param max Maximum number
     * @return returns a random number between the minimum and maximum value
     */
    private int next(int min, int max)
    {
        // java random code copied from https://www.educative.io/edpresso/how-to-generate-random-numbers-in-java
        return (int) Math.floor(Math.random() * (max - min) + min);
    }

}

File: BlackJack-CSHARP/Blackjack21.Game/Game/Helper/ShuffleList.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography;

namespace Blackjack21.Game.Helper
{
    /// <summary>
    /// A class used to shuffle a list's objects
    /// </summary>
    /// <typeparam name="T">The object to use for the list</typeparam>
    public class ShuffleList<T> : List<T>
    {
        private Random _random;

        /// <summary>
        /// Initialize the shuffle list object
        /// </summary>
        public ShuffleList()
        {
            var cryptoResult = new byte[4];
            new RNGCryptoServiceProvider().GetBytes(cryptoResult);
            int seed = BitConverter.ToInt32(cryptoResult, 0);
            _random = new Random(seed);
        }

        /// <summary>
        /// Add a new item to the list
        /// </summary>
        /// <param name="item">Item to add to the list</param>
        /// <param name="isRandom">Add the item to a random location</param>
        public void Add(T item, bool isRandom = false)
        {
            if (Count == 0 || isRandom == false)
            {
                base.Add(item);
            }
            else
            {
                int index = Next(0, Count + 1);
                if (index == Count)
                {
                    base.Add(item);
                }
                else
                {
                    T x = this[index];
                    this[index] = item;
                    base.Add(x);
                }
            }
        }

        /// <summary>
        /// Shuffle items in the list by removing each item randomly and adding it to a seperate list before rebuilding the final list
        /// </summary>
        public void ShuffleShift()
        {
            List<T> tmp = new List<T>();

            while (this.Count > 0)
            {
                int i = Next(0, base.Count);
                tmp.Add(base[i]);
                base.RemoveAt(i);
            }

            base.AddRange(tmp.ToArray());

            tmp.Clear();
        }

        /// <summary>
        /// Shuffle items inplace by swapping locations randomly
        /// </summary>
        public void ShuffleInplace()
        {
            _random = new Random();
            for (int i = base.Count - 1; i >= 0; i--)
            {
                T tmp = base[i];
                int randomIndex = Next(0, i + 1);

                //Swap elements
                base[i] = base[randomIndex];
                base[randomIndex] = tmp;
            }
        }

        /// <summary>
        /// Removes the first item on top of the list and returns it
        /// </summary>
        /// <returns>Returns the retmoved item</returns>
        public T Pop()
        {
            T t = base[0];
            base.RemoveAt(0);
            return t;
        }

        /// <summary>
        /// Removes a random item from the list and returns it
        /// </summary>
        /// <returns>Returns the removed list item</returns>
        public T PopRandom()
        {
            int i = Next(0, base.Count);
            T t = base[i];
            base.RemoveAt(i);
            return t;
        }

        /// <summary>
        /// Selects a random item from the list without removing it
        /// </summary>
        /// <returns>Returns the list item</returns>
        public T SelectRandom()
        {
            int i = Next(0, base.Count);
            T t = base[i];
            return t;
        }

        /// <summary>
        /// Select the next random number
        /// </summary>
        /// <param name="min">Minimum number</param>
        /// <param name="max">Maximum number</param>
        /// <returns>returns a random number between the minimum and maximum value</returns>
        private int Next(int min, int max)
        {
            return _random.Next(min, max);
        }
    }
}
